crosslink/internal/crosslink.go

```diff
@@ -0,0 +1,129 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crosslink
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/mod/modfile"
+)
+
+func Crosslink(rc runConfig) {
+	var err error
+
+	rootModulePath, err := identifyRootModule(rc.RootPath)
+	if err != nil {
+		panic(fmt.Sprintf("failed to identify root module: %v", err))
+	}
+
+	graph, err := buildDepedencyGraph(rc, rootModulePath)
+	if err != nil {
+		panic(fmt.Sprintf("failed to build dependency graph: %v", err))
+	}
+
+	for _, moduleInfo := range graph {
+		err = insertReplace(&moduleInfo, rc)
+		if err != nil {
+			panic(fmt.Sprintf("failed to insert replace statements: %v", err))
+		}
+
+		err = pruneReplace(rootModulePath, &moduleInfo, rc)
+
+		if err != nil {
+			panic(fmt.Sprintf("error pruning replace statements: %v", err))
+		}
+
+		err = writeModules(moduleInfo)
+		if err != nil {
+			panic(fmt.Sprintf("error writing gomod files: %v", err))
+		}
+	}
+	err = rc.logger.Sync()
+	if err != nil {
+		fmt.Printf("failed to sync logger:  %v", err)
+	}
+}
+
+func insertReplace(module *moduleInfo, rc runConfig) error {
+	// modfile type that we will work with then write to the mod file in the end
+	mfParsed, err := modfile.Parse("gomod", module.moduleContents, nil)
+	if err != nil {
+		return err
+	}
+
+	for reqModule := range module.requiredReplaceStatements {
+		// skip excluded
+		if _, exists := rc.ExcludedPaths[reqModule]; exists {
+			if rc.Verbose {
+				rc.logger.Sugar().Infof("Excluded Module %s, ignoring replace", reqModule)
```

Comment:
```suggestion
				rc.logger.Info("Excluded Module encountered, ignoring replace", zap.String("module", reqModule))
```

I would probably prefer to use a structured logging style throughout, rather than intermixing variable information into the log messages.